### PR TITLE
guest_os_booting: add new case for ovmf loader

### PR DESCRIPTION
**Pull request by meinaLi**
This PR automates:
VIRT-296930 - [OVMF] Boot vm with ovmf loader

### CODE REVIEW

**Comment by meinaLi** [comment]
```
# avocado run --vt-type libvirt --test-runner=runner guest_os_booting.ovmf_loader
JOB ID     : 5a76eecb36cc94a07114dbad1dbb7d0a9d4ed867
JOB LOG    : /var/lib/avocado/job-results/job-2023-05-19T07.23-5a76eec/job.log
 (1/6) type_specific.io-github-autotest-libvirt.guest_os_booting.ovmf_loader.positive_test.common_value: PASS (78.05 s)
 (2/6) type_specific.io-github-autotest-libvirt.guest_os_booting.ovmf_loader.positive_test.secure_no: PASS (79.60 s)
 (3/6) type_specific.io-github-autotest-libvirt.guest_os_booting.ovmf_loader.positive_test.stateless_yes: PASS (21.42 s)
 (4/6) type_specific.io-github-autotest-libvirt.guest_os_booting.ovmf_loader.negative_test.readonly_no: PASS (18.86 s)
 (5/6) type_specific.io-github-autotest-libvirt.guest_os_booting.ovmf_loader.negative_test.type_rom: PASS (18.04 s)
 (6/6) type_specific.io-github-autotest-libvirt.guest_os_booting.ovmf_loader.negative_test.incorrect_loader_path: PASS (17.62 s)
RESULTS    : PASS 6 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /var/lib/avocado/job-results/job-2023-05-19T07.23-5a76eec/results.html
JOB TIME   : 235.31 s
```

**Comment by meinaLi** [comment]
Label: modular_redesign

**Comment by meinaLi** [comment]
Depend on https://github.com/avocado-framework/avocado-vt/pull/3691

**Comment by meinaLi** [comment]
@dzhengfy @Yingshun @brucepatricklee
This is the first modular PR in guest os booting feature. Could you help to review it earlier? In this way, I can avoid some serious problems that do not conform to the format when I write the following case. Thanks.

File: libvirt/tests/cfg/guest_os_booting/ovmf_firmware/ovmf_loader.cfg [hunk 1]
```diff
@@ -0,0 +1,35 @@
+- guest_os_booting.ovmf_loader:
+    type = ovmf_loader
+    start_vm = no
+    status_error = "no"
+    test_cmd = "dmesg | grep -i secure"
+    loader_path = "/usr/share/edk2/ovmf/OVMF_CODE.secboot.fd"
+    os_type = {'type': 'hvm', 'machine': 'q35'}
```
Comment (Yingshun) [inline libvirt/tests/cfg/guest_os_booting/ovmf_firmware/ovmf_loader.cfg:7]: is this case only for q35?

File: libvirt/tests/cfg/guest_os_booting/ovmf_firmware/ovmf_loader.cfg [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+        - positive_test:
+            variants:
+                - common_value:
+                    loader_dict = {'secure': 'yes', 'loader_readonly': 'yes', 'loader_type': 'pflash', 'loader': '${loader_path}'}
+                - secure_no:
+                    nvram_path = "/var/lib/libvirt/qemu/nvram/nvram_VARS.fd"
+                    nvram_template = "/usr/share/edk2/ovmf/OVMF_VARS.fd"
+                    loader_dict = {'loader': '${loader_path}', 'nvram': '${nvram_path}', 'nvram_attrs': {'template': '${nvram_template}'}, 'secure': 'no', 'loader_readonly': 'yes', 'loader_type': 'pflash'}
+                - stateless_yes:
+                    special_case = "yes"
```
Comment (Yingshun) [inline libvirt/tests/cfg/guest_os_booting/ovmf_firmware/ovmf_loader.cfg:19]: Can you please give a proper name? I think i don't understand what kind of cases are special ones here.

File: provider/guest_os_booting/guest_os_booting_base.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
+import logging as log
+
+from virttest import remote
+from virttest import virsh
+from virttest.libvirt_xml import vm_xml
+from virttest.utils_libvirt import libvirt_bios
+
+
+# Using as lower captical is not the best way to do, but this is just a workaround
+# to avoid changing the entire file.
```
Comment (Yingshun) [inline provider/guest_os_booting/guest_os_booting_base.py:10]: These two lines were added to describe the situation we faced before. so please delete them.

File: provider/guest_os_booting/guest_os_booting_base.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    feature_xml = vmxml.features
+    feature_xml.smm = smm_state
+    if smm_size:
+        feature_xml.smm_tseg = smm_size
+    vmxml.features = feature_xml
+    vmxml.sync()
+    return vmxml
+
+
+def start_boot_guest(vm, vm_name, test_cmd, params):
```
Comment (Yingshun) [inline provider/guest_os_booting/guest_os_booting_base.py:48]: May I ask what the difference between 'start' and 'boot' is?

File: provider/guest_os_booting/guest_os_booting_base.py [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+    """
+    Start and boot the guest
+    :params vm: vm object
+    :params vm_name: the name of guest
+    :params test_cmd: command run in guest to check it works
+    """
+    username = params.get("username", "root")
+    password = params.get("password", "redhat")
```
Comment (Yingshun) [inline provider/guest_os_booting/guest_os_booting_base.py:56]: I think we should avoid texting plain password in the code, can you remove it? what do you think?

File: provider/guest_os_booting/guest_os_booting_base.py [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+    if not vm.is_alive():
+        virsh.start(vm_name, "--reset-nvram", timeout=30, debug=True)
+    vm_ip = vm.wait_for_get_address(0, timeout=240)
+    remote_session = remote.wait_for_login("ssh", vm_ip, "22", username, password,
```
Comment (Yingshun) [inline provider/guest_os_booting/guest_os_booting_base.py:60]: What about using vm.wait_for_login() to get vm's session from 'vm' object?

File: provider/guest_os_booting/guest_os_booting_base.py [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+
+# Using as lower captical is not the best way to do, but this is just a workaround
+# to avoid changing the entire file.
+LOG = log.getLogger('avocado.' + __name__)
+
+
+def prepare_os_xml(vm_name, os_dict):
+    """
+    Prepare a guest with related os loader xml.
+    :params vm_name: the name of guest
```
Comment (dzhengfy) [inline provider/guest_os_booting/guest_os_booting_base.py:17]: Add blank line above this line, please.

File: provider/guest_os_booting/guest_os_booting_base.py [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+    osxml.setup_attrs(**os_dict)
+    vmxml.os = osxml
+    vmxml.sync()
+    return vmxml
+
+
+def prepare_smm_xml(vm_name, smm_state, smm_size):
+    """
+    Prepare a guest with related feature smm xml.
+    :params vm_name: the name of guest
```
Comment (dzhengfy) [inline provider/guest_os_booting/guest_os_booting_base.py:33]: Add blank line above this line, please.

File: provider/guest_os_booting/guest_os_booting_base.py [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+        feature_xml.smm_tseg = smm_size
+    vmxml.features = feature_xml
+    vmxml.sync()
+    return vmxml
+
+
+def start_boot_guest(vm, vm_name, test_cmd, params):
+    """
+    Start and boot the guest
+    :params vm: vm object
```
Comment (dzhengfy) [inline provider/guest_os_booting/guest_os_booting_base.py:51]: Add blank line above this line, please.

File: provider/guest_os_booting/guest_os_booting_base.py (hunk 5 above)
Comment (dzhengfy) [reply provider/guest_os_booting/guest_os_booting_base.py:56]: For line 55 and 56, we will use CI to replace them with real value. We do not add credential information here. To access the vm, we could reuse existing variable in virttest/shared/cfg/guest-os/Linux.cfg

File: libvirt/tests/cfg/guest_os_booting/ovmf_firmware/ovmf_loader.cfg (hunk 1 above)
Comment (dzhengfy) [inline libvirt/tests/cfg/guest_os_booting/ovmf_firmware/ovmf_loader.cfg:7]: I suggest we also consider aarch64 at least here.

File: provider/guest_os_booting/guest_os_booting_base.py (hunk 6 above)
Comment (dzhengfy) [reply provider/guest_os_booting/guest_os_booting_base.py:60]: Is it a must to get the vm ip? If yes, we need to add a checkpoint here. Otherwise, we could use vm.wait_for_login() ?

File: provider/guest_os_booting/guest_os_booting_base.py [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+    password = params.get("password", "redhat")
+    if not vm.is_alive():
+        virsh.start(vm_name, "--reset-nvram", timeout=30, debug=True)
+    vm_ip = vm.wait_for_get_address(0, timeout=240)
+    remote_session = remote.wait_for_login("ssh", vm_ip, "22", username, password,
+                                           r"[\#\$]\s*$")
+    status, output = remote_session.cmd_status_output(test_cmd)
+    LOG.debug("CMD '%s' running result is:\n%s", test_cmd, output)
+    if status:
+        LOG.fail("Failed to boot %s", vm_name)
```
Comment (dzhengfy) [inline provider/guest_os_booting/guest_os_booting_base.py:65]: `test.fail()`

**Comment by meinaLi** [comment]
```
# avocado run --vt-type libvirt --test-runner=runner guest_os_booting.ovmf_loader
JOB ID     : dd6dd2bea33884bd432225057668ac7dffdf4414
JOB LOG    : /var/lib/avocado/job-results/job-2023-05-30T03.11-dd6dd2b/job.log
 (1/6) type_specific.io-github-autotest-libvirt.guest_os_booting.ovmf_loader.positive_test.default_value: PASS (79.33 s)
 (2/6) type_specific.io-github-autotest-libvirt.guest_os_booting.ovmf_loader.positive_test.secure_no: PASS (90.38 s)
 (3/6) type_specific.io-github-autotest-libvirt.guest_os_booting.ovmf_loader.positive_test.stateless_yes: PASS (16.19 s)
 (4/6) type_specific.io-github-autotest-libvirt.guest_os_booting.ovmf_loader.negative_test.readonly_no: PASS (19.64 s)
 (5/6) type_specific.io-github-autotest-libvirt.guest_os_booting.ovmf_loader.negative_test.type_rom: PASS (18.73 s)
 (6/6) type_specific.io-github-autotest-libvirt.guest_os_booting.ovmf_loader.negative_test.incorrect_loader_path: PASS (19.50 s)
RESULTS    : PASS 6 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /var/lib/avocado/job-results/job-2023-05-30T03.11-dd6dd2b/results.html
JOB TIME   : 245.54 s
```

File: libvirt/tests/cfg/guest_os_booting/ovmf_firmware/ovmf_loader.cfg (hunk 1 above)
Comment (meinaLi) [reply libvirt/tests/cfg/guest_os_booting/ovmf_firmware/ovmf_loader.cfg:6]: Yes, it is.

File: libvirt/tests/cfg/guest_os_booting/ovmf_firmware/ovmf_loader.cfg (hunk 2 above)
Comment (meinaLi) [reply libvirt/tests/cfg/guest_os_booting/ovmf_firmware/ovmf_loader.cfg]: Updated it to stateless = "yes"

File: provider/guest_os_booting/guest_os_booting_base.py (hunk 3 above)
Comment (meinaLi) [reply provider/guest_os_booting/guest_os_booting_base.py]: Have already removed.

File: provider/guest_os_booting/guest_os_booting_base.py (hunk 4 above)
Comment (meinaLi) [reply provider/guest_os_booting/guest_os_booting_base.py]: Here 'start' only means `virsh start` success, 'boot' means the guest can be accessed.

File: provider/guest_os_booting/guest_os_booting_base.py (hunk 5 above)
Comment (meinaLi) [reply provider/guest_os_booting/guest_os_booting_base.py]: Thanks for your useful information. I changed to use vm.wait_for_login(). So need the username and password again.

File: provider/guest_os_booting/guest_os_booting_base.py (hunk 6 above)
Comment (meinaLi) [reply provider/guest_os_booting/guest_os_booting_base.py]: Updated.

File: libvirt/tests/cfg/guest_os_booting/ovmf_firmware/ovmf_loader.cfg [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+    status_error = "no"
+    test_cmd = "dmesg | grep -i secure"
+    loader_path = "/usr/share/edk2/ovmf/OVMF_CODE.secboot.fd"
+    os_type = {'type': 'hvm', 'machine': 'q35'}
+    smm_state = "on"
+    variants:
+        - positive_test:
+            variants:
+                - common_value:
+                    loader_dict = {'secure': 'yes', 'loader_readonly': 'yes', 'loader_type': 'pflash', 'loader': '${loader_path}'}
```
[the remark this answers is not in the document]
Comment (meinaLi) [reply libvirt/tests/cfg/guest_os_booting/ovmf_firmware/ovmf_loader.cfg:12]: Yes, it's default values.

File: libvirt/tests/src/guest_os_booting/ovmf_firmware/ovmf_loader.py [hunk 12]
```diff
@@ -0,0 +1,51 @@
+import os
```
[the remark this answers is not in the document]
Comment (meinaLi) [reply libvirt/tests/src/guest_os_booting/ovmf_firmware/ovmf_loader.py:11]: Add the copyright. But I get 
```
libvirt/tests/src/guest_os_booting/ovmf_firmware/ovmf_loader.py:8:0: C0401: Wrong spelling of a word 'meili' in a comment:
#   Author: Meina Li <meili@redhat.com>
                      ^^^^^
Did you mean: ''milieu' or 'mem' or 'mins' or 'multi''? (wrong-spelling-in-comment)
```
That means I always get the new python file failed when spellcheck. Do we have some valid methods to avoid this?

File: libvirt/tests/src/guest_os_booting/ovmf_firmware/ovmf_loader.py [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+    bkxml = vmxml.copy()
+
+    try:
+        guest_os.prepare_smm_xml(vm_name, smm_state, "")
+        guest_os.prepare_os_xml(vm_name, os_dict)
+        virsh.dumpxml(vm_name, debug=True)
+        if special_case:
+            virsh.start(vm_name, debug=True)
+            if "OVMF.amdsev.fd" in virsh.dumpxml(vm_name).stdout_text:
+                test.log.debug("The stateless guest use correct loader automatically")
```
[the remark this answers is not in the document]
Comment (meinaLi) [reply libvirt/tests/src/guest_os_booting/ovmf_firmware/ovmf_loader.py:46]: Added the test.fail(). Thanks.

File: libvirt/tests/src/guest_os_booting/ovmf_firmware/ovmf_loader.py [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+        if not status_error and not special_case:
+            guest_os.start_boot_guest(vm, vm_name, test_cmd, params)
+            virsh.dumpxml(vm_name, debug=True)
+        if status_error:
+            if use_file:
+                fd = os.open(incorrect_loader_path, os.O_CREAT)
+                os.close(fd)
+            ret = virsh.start(vm_name)
+            if ret.exit_status:
```
[the remark this answers is not in the document]
Comment (meinaLi) [reply libvirt/tests/src/guest_os_booting/ovmf_firmware/ovmf_loader.py:56]: Here we need to use this part to test the negative cases in cfg.

File: provider/guest_os_booting/guest_os_booting_base.py (hunk 8 above)
Comment (meinaLi) [reply provider/guest_os_booting/guest_os_booting_base.py:32]: Updated

File: provider/guest_os_booting/guest_os_booting_base.py (hunk 9 above)
Comment (meinaLi) [reply provider/guest_os_booting/guest_os_booting_base.py:51]: Updated

File: libvirt/tests/cfg/guest_os_booting/ovmf_firmware/ovmf_loader.cfg (hunk 1 above)
Comment (meinaLi) [reply libvirt/tests/cfg/guest_os_booting/ovmf_firmware/ovmf_loader.cfg:6]: I think the arch attribute in <type arch='x86_64' machine='pc-q35-rhel9.0.0'>hvm</type> only show the different arches. Here I didn't use it.

File: provider/guest_os_booting/guest_os_booting_base.py (hunk 10 above)
Comment (meinaLi) [reply provider/guest_os_booting/guest_os_booting_base.py]: Updated.

File: libvirt/tests/src/guest_os_booting/ovmf_firmware/ovmf_loader.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (dzhengfy) [reply libvirt/tests/src/guest_os_booting/ovmf_firmware/ovmf_loader.py:11]: Add your id into spell.ignore file same as we did

File: provider/guest_os_booting/guest_os_booting_base.py [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+def start_guest_failed(vm, vm_name, error_msg, test):
+    """
+    Start guest failed for some negative test scenarios
+
+    :params vm: vm object
+    :params vm_name: the name of guest
+    """
+    ret = virsh.start(vm_name, "--reset-nvram", timeout=30, debug=True)
+    if ret.exit_status:
+        libvirt.check_result(ret, expected_fails=error_msg)
```
Comment (dzhengfy) [inline provider/guest_os_booting/guest_os_booting_base.py:80]: This libvirt.check_result() can help check` ret.exit_status`, so you really need not to check it by yourself.

File: libvirt/tests/src/guest_os_booting/ovmf_firmware/ovmf_loader.py [hunk 16]
```diff
@@ -0,0 +1,58 @@
...
+                test.log.debug("The stateless guest use correct loader automatically")
+            else:
+                test.fail("Can't get expected loader path for stateless")
+        if not status_error and not stateless:
+            guest_os.start_boot_guest(vm, vm_name)
+        if status_error:
+            if use_file:
+                fd = os.open(incorrect_loader_path, os.O_CREAT)
+                os.close(fd)
+            guest_os.start_guest_failed(vm, vm_name, error_msg, test)
```
Comment (dzhengfy) [inline libvirt/tests/src/guest_os_booting/ovmf_firmware/ovmf_loader.py:54]: To me, start_guest_failed() and start_boot_guest() are a little redundunt because their codes could be merged together, like below:
```

if not vm.is_alive():
        virsh.start(vm_name, "--reset-nvram", timeout=30, debug=True)
virsh.dumpxml(vm_name, debug=True)
libvirt.check_result(ret, expected_fails=error_msg)
if not error_msg:
    vm.wait_for_login().close()
    LOG.debug("Succeed to boot %s", vm_name)


```

File: libvirt/tests/src/guest_os_booting/ovmf_firmware/ovmf_loader.py [hunk 17]
```diff
@@ -0,0 +1,58 @@
...
+    try:
+        guest_os.prepare_smm_xml(vm_name, smm_state, "")
+        guest_os.prepare_os_xml(vm_name, loader_dict)
+        virsh.dumpxml(vm_name, debug=True)
+        if stateless:
+            virsh.start(vm_name, debug=True)
+            if "OVMF.amdsev.fd" in virsh.dumpxml(vm_name).stdout_text:
+                test.log.debug("The stateless guest use correct loader automatically")
+            else:
+                test.fail("Can't get expected loader path for stateless")
```
[the remark this answers is not in the document]
Comment (meinaLi) [reply libvirt/tests/src/guest_os_booting/ovmf_firmware/ovmf_loader.py]: Could you check it again? I'm not sure if it's clear now.

File: provider/guest_os_booting/guest_os_booting_base.py (hunk 15 above)
Comment (meinaLi) [reply provider/guest_os_booting/guest_os_booting_base.py]: Thanks for your suggestion. I've updated it.

File: libvirt/tests/src/guest_os_booting/ovmf_firmware/ovmf_loader.py (hunk 16 above)
Comment (meinaLi) [reply libvirt/tests/src/guest_os_booting/ovmf_firmware/ovmf_loader.py]: Merged them together.

**Review by dzhengfy — APPROVED** [review]
LGTM

File: libvirt/tests/cfg/guest_os_booting/ovmf_firmware/ovmf_loader.cfg (hunk 1 above)
Comment (Yingshun) [reply libvirt/tests/cfg/guest_os_booting/ovmf_firmware/ovmf_loader.cfg]: do you need to add 'only q35' in cfg?

File: provider/guest_os_booting/guest_os_booting_base.py (hunk 4 above)
Comment (Yingshun) [reply provider/guest_os_booting/guest_os_booting_base.py]: Does 'boot' mean that guest is fully started up? From the code, if error_msg is not None, it'll not access the guess. if so, can you rename it? what do you think of  'check_vm_startup'?

File: libvirt/tests/src/guest_os_booting/ovmf_firmware/ovmf_loader.py [hunk 18]
```diff
@@ -0,0 +1,55 @@
...
+    stateless = "yes" == params.get("stateless", "no")
+
+    vm = env.get_vm(vm_name)
+    vmxml = vm_xml.VMXML.new_from_inactive_dumpxml(vm_name)
+    bkxml = vmxml.copy()
+
+    try:
+        guest_os.prepare_smm_xml(vm_name, smm_state, "")
+        guest_os.prepare_os_xml(vm_name, loader_dict)
+        # stateless='yes' only use for AMD test, so here we only check the dumpxml for it to avoid the machine issue
```
[the remark this answers is not in the document]
Comment (meinaLi) [reply libvirt/tests/src/guest_os_booting/ovmf_firmware/ovmf_loader.py:40]: No, because despite this feature is only used for AMD, but we can also start the guest in other arches. So here I did a special check which is different from other test scenarios: only check the dumpxml is as expected.

File: provider/guest_os_booting/guest_os_booting_base.py [hunk 19]
```diff
@@ -0,0 +1,71 @@
...
+    vmxml = vm_xml.VMXML.new_from_inactive_dumpxml(vm_name)
+    vmxml.os = libvirt_bios.remove_bootconfig_items_from_vmos(vmxml.os)
+    LOG.debug("Set the os xml")
+    vmxml.setup_attrs(os=os_dict)
+    # Some negative cases may need to get expected error when defining guest
+    try:
+        vmxml.sync()
+    except xcepts.LibvirtXMLError as details:
+        if "Failed to define" in str(details):
+            raise ValueError("%s" % str(details).split(':')[-1])
```
[the remark this answers is not in the document]
Comment (meinaLi) [reply provider/guest_os_booting/guest_os_booting_base.py:29]: This part is from other test cases: https://github.com/autotest/tp-libvirt/pull/4924. Also in some other test cases which are still not automated, they may also have the same negative test scenarios: define the guest failed with invalid values. Because this PR has not already merged and I've already drafted PR4924, so I directly updated it to this PR.

**Comment by meinaLi** [comment]
For the function start_boot_guest(), I've updated it to check_vm_startup(). The previous one is really not correct after merging both test scenarios.

File: provider/guest_os_booting/guest_os_booting_base.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (meinaLi) [reply provider/guest_os_booting/guest_os_booting_base.py]: Updated.

File: libvirt/tests/src/guest_os_booting/ovmf_firmware/ovmf_loader.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (meinaLi) [reply libvirt/tests/src/guest_os_booting/ovmf_firmware/ovmf_loader.py:42]: Use the libvirt_vmxml.check_guest_xml_by_xpaths function to check the dumpxml.

File: libvirt/tests/cfg/guest_os_booting/ovmf_firmware/ovmf_loader.cfg (hunk 1 above)
Comment (Yingshun) [reply libvirt/tests/cfg/guest_os_booting/ovmf_firmware/ovmf_loader.cfg]: any comments?

Comment (meinaLi) [reply libvirt/tests/cfg/guest_os_booting/ovmf_firmware/ovmf_loader.cfg]: Sorry for forgetting this point. For safety, I've added "only q35" to cfg.

**Review by Yingshun — APPROVED** [review]
